sklearn/cluster/_hdbscan/_boruvka.pyx

```diff
@@ -0,0 +1,760 @@
+        NodeData_t[::1] node_data
+        BoruvkaUnionFind component_union_find
+        MST_edge_t[::1] edges
+        intp_t num_edges
+
+    def __init__(
+        self,
+        tree,
+        min_samples=5,
+        metric='euclidean',
+        leaf_size=20,
+        alpha=1.0,
+        approx_min_span_tree=False,
+        n_jobs=None,
+        **kwargs
+    ):
+
+        self.tree =tree
+        self.has_KDTree = isinstance(tree, KDTree)
+        self.raw_data = self.tree.data
+        self.node_bounds = self.tree.node_bounds
+        self.alpha = alpha
+        self.approx_min_span_tree = approx_min_span_tree
+        self.n_jobs = effective_n_jobs(n_jobs)
+        self.min_samples = min_samples
+
+        self.num_points = self.tree.data.shape[0]
+        self.num_features = self.tree.data.shape[1]
+        self.num_nodes = self.tree.node_data.shape[0]
+
+        self.dist = DistanceMetric.get_metric(metric, **kwargs)
+
+        self.components = np.arange(self.num_points, dtype=np.intp)
+        self.bounds = np.empty(self.num_nodes, np.float64)
+        self.component_of_point = np.empty(self.num_points, dtype=np.intp)
+        self.component_of_node = np.empty(self.num_nodes, dtype=np.intp)
+        self.candidate_neighbor = np.empty(self.num_points, dtype=np.intp)
+        self.candidate_point = np.empty(self.num_points, dtype=np.intp)
+        self.candidate_distance = np.empty(self.num_points, dtype=np.float64)
+        self.component_union_find = BoruvkaUnionFind(self.num_points)
+
+        self.edges = np.empty((self.num_points - 1,), dtype=MST_edge_dtype)
+        self.num_edges = 0
+
+        self.idx_array = self.tree.idx_array
+        self.node_data = self.tree.node_data
+
+        if not self.has_KDTree:
+            # Compute centroids for BallTree
+            self.centroid_distances = self.dist.pairwise(self.tree.node_bounds[0])
+
+        self._initialize_components()
+        self._compute_bounds()
+
+    cdef _compute_bounds(self):
+        """Initialize core distances"""
+
+        cdef intp_t i, n, m
+
+        cdef cnp.ndarray[float64_t, ndim=2] knn_dist
+        cdef cnp.ndarray[intp_t, ndim=2] knn_indices
+
+        # TODO: Evaluate query-parallelization featured in original HDBSCAN
+        # implementation. Removed for now for simplicity.
+        knn_dist, knn_indices = self.tree.query(
+            self.tree.data,
+            k=self.min_samples,
+            dualtree=True,
+            breadth_first=True
+        )
+
+        self.core_distance = knn_dist[:, self.min_samples - 1].copy()
+
+        # Since we already computed NN distances for the min_samples closest
+        # points we can use this to do the first round of boruvka -- we won't
+        # get every point due to core_distance/mutual reachability distance
+        # issues, but we'll get quite a few, and they are the hard ones to
+        # get, so fill in any we can and then run update components.
+        for n in range(self.num_points):
```

Comment:
This loop iterations are independent, hence they can be run in parallel if needed.